File: capacity-club-api/src/domain-modules/indoor-training/model/entity/indoor-training.entity.ts
import { BaseEntity } from '@common/model';
import { UniqueId, uniqueIdTransformer } from '@common/model/unique-id';
import { ApiProperty } from '@nestjs/swagger';
import { Transform } from 'class-transformer';
import { IndoorTrainingSubscription } from 'domain-modules/indoor-training-subscription/model/entity/indoor-training-subscription.entity';
import { Workout } from 'domain-modules/workout/model/entity/workout.entity';
import {
  Column,
  Entity,
  JoinColumn,
  OneToMany,
  OneToOne,
  PrimaryColumn,
} from 'typeorm';
/**
 * Entity representing an indoor training session.
 * These training sessions can be individual or collective.
 */
@Entity()
export class IndoorTraining extends BaseEntity {
  /**
   * Unique identifier for the indoor training.
   * Uses a custom transformer for serialization.
   */
  @PrimaryColumn('varchar')
  @Transform(uniqueIdTransformer.to, { toClassOnly: true })
  @Transform(uniqueIdTransformer.from, { toPlainOnly: true })
  indoor_training_id: UniqueId;
  @Column({ nullable: true })
  title: string;
  @Column({ nullable: false })
  training_date: Date;
  @Column({ nullable: false })
  start_hours: string;
  @Column({ nullable: true })
  end_hours: string;
  @Column({ nullable: false })
  nb_place: number;
  @Column({ nullable: false })
  nb_subscription: number;
  @Column({ nullable: false })
  is_collective: boolean;
  @Column({ nullable: true })
  is_active: boolean;

  @OneToOne(() => Workout, {
    lazy: true,
    nullable: true,
  })
  @JoinColumn({ name: 'workout_id', referencedColumnName: 'workout_id' })
  @ApiProperty({ type: () => Workout })
  workout: Promise<Workout>;
  @OneToMany(() => IndoorTrainingSubscription, (data) => data.indoor_training, {
    cascade: true,
    lazy: true,
    nullable: true,
  })
  @ApiProperty({ type: () => IndoorTrainingSubscription, isArray: true })
  indoor_training_subscription_list: Promise<IndoorTrainingSubscription[]>;
}

File: capacity_club_mobile_app/lib/domain-features/domain/indoor-training/entity/indoor_training_entity.dart
import 'package:capacity_club_mobile_app/core/model/entities/unique_id.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/indoor-training-subscription/entity/indoor_training_subcription_entity.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/workout/entity/workout_entity.dart';

class IndoorTrainingEntity {
  final UniqueId indoor_training_id;
  final String title;
  final DateTime training_date;
  final String start_hours;
  final String end_hours;
  final int nb_place;
  final int nb_subscription;
  final bool is_collective;
  final bool is_active;
  final WorkoutEntity workout;
  final List<IndoorTrainingSubcriptionEntity> indoor_training_subscription_list;
  final bool is_user_registred;
  final bool is_empty;

  IndoorTrainingEntity(
      {required this.indoor_training_id,
      required this.title,
      required this.training_date,
      required this.start_hours,
      required this.end_hours,
      required this.nb_place,
      required this.nb_subscription,
      required this.is_collective,
      required this.is_active,
      required this.workout,
      required this.indoor_training_subscription_list,
      this.is_user_registred = false,
      this.is_empty = false});

  factory IndoorTrainingEntity.empty() {
    return IndoorTrainingEntity(
        indoor_training_id: UniqueId(''),
        title: '',
        training_date: DateTime.now(),
        start_hours: '',
        end_hours: '',
        nb_place: 0,
        nb_subscription: 0,
        is_collective: false,
        is_active: false,
        is_user_registred: false,
        workout: WorkoutEntity.empty(),
        indoor_training_subscription_list: [],
        is_empty: true);
  }
}
